ClientSyncComponent/ClientSyncComponent.Client/Pages/ClientSyncComponent.razor.cs

```diff
@@ -0,0 +1,170 @@
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.JSInterop;
+
+namespace ClientSyncComponent.Client.Pages
+{
+    /// <summary>
+    /// Component that integrates with puzzle.js to sync changes to Azure Tables
+    /// </summary>
+    public partial class ClientSyncComponent
+    {
+        [Inject]
+        IJSRuntime JSRuntime { get; set; }
+
+        [Parameter]
+        public int PuzzleId { get; set; }
+
+        [Parameter]
+        public int TeamId { get; set; }
+
+        [Parameter]
+        public int PuzzleUserId { get; set; }
+
+        [Parameter]
+        public Uri TableSASUrl { get; set; }
+
+        DateTimeOffset LastSyncUtc = DateTimeOffset.MinValue;
+
+        TableClient TableClient { get; set; }
+
+        List<string> ReceivedValues { get; set; } = new List<string>();
+
+        Timer Timer { get; set; }
+
+        protected override Task OnParametersSetAsync()
+        {
+            TableClient = new TableClient(TableSASUrl);
+            return base.OnParametersSetAsync();
+        }
+
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            if (firstRender)
+            {
+                await JSRuntime.InvokeVoidAsync("registerDotNet", DotNetObjectReference.Create(this));
+            }
+
+            await base.OnAfterRenderAsync(firstRender);
+        }
+
+        private async void OnTimer(object? state)
+        {
+            await SyncAsync();
+            string visibility = await JSRuntime.InvokeAsync<string>("getVisibility");
+            Console.WriteLine("Timer ticked. Visibility is " + visibility);
+
+            int newInterval = CalculateInterval();
+            Timer.Change(newInterval, newInterval);
+        }
+
+        /// <summary>
+        /// Heuristic to minimize load depending on how active a puzzle is
+        /// </summary>
+        /// <returns>Interval in milliseconds to poll storage at</returns>
+        private int CalculateInterval()
+        {
+            // todo: get a multiplier or override from the server for central updates?
+            TimeSpan timeSinceUpdate = DateTimeOffset.UtcNow - LastSyncUtc;
+            if (timeSinceUpdate.TotalSeconds < 120)
+            {
+                return 1000;
+            }
+            else if (timeSinceUpdate.TotalMinutes < 10)
+            {
+                return 2000;
+            }
+            else if (timeSinceUpdate.TotalHours < 1)
+            {
+                return 5000;
+            }
+            else if (timeSinceUpdate.TotalHours < 6)
+            {
+                return 10000;
+            }
+            else
+            {
+                return 60000;
+            }
+        }
+
+        [JSInvokable]
+        public void VisibilityChanged(string visibility)
+        {
+            Console.WriteLine($".NET Visibility changed to {visibility}");
+            if (visibility == "visible")
+            {
+                Timer.Change(0, CalculateInterval());
```

Comment:
This should be the min value that `CalculateInterval` can return given current server conditions - if I join co-op after 6 hours solo, it should not take a minute to sync.

Reply:
The first parameter is the duetime of the first call, it'll recalculate the interval after that, so there should be no delay

Reply:
OK but this is also the "demo bug" you described today

Reply:
Discussed in person - no it isn't the "demo bug". Makes sense now.